File: dotnet/01-commands/explain/LifeGame/GameTest.cs
using Xunit;

namespace LifeGame
{
    public class GameTest
    {
        [Fact]
        public void GivenAnGridWithAllCellsDead_TheGameIsFinishedAfterThFirstRound()
        {

            var grid = Grid.TryCreate(1, 1, new List<Cell> { });
            var game = new Game(grid);

            game.Run();

            Assert.True(game.IsFinished);
        }

        [Fact]
        public void GivenInitialState_ThenNotFinished()
        {
            var grid = Grid.TryCreate(1, 1, new List<Cell> { });
            var game = new Game(grid);

            Assert.False(game.IsFinished);

        }

        [Fact]
        public void GivenA1xA1GridWith1Cell_GameIsNotFinishedAfterOneRound()
        {
            var cell = new Cell(1,1);
            var grid = Grid.TryCreate(1, 1, new List<Cell> { cell });

            var game = new Game(grid);

            game.Run();

            Assert.False(game.IsFinished);
        }

        [Fact]
        public void GivenA1xA1GridWithCell_GameIsFinishedAfterTwoRounds()
        {
            var cell = new Cell(1,1);
            var grid = Grid.TryCreate(1, 1, new List<Cell> { cell });

            var game = new Game(grid);

            game.Run();
            game.Run();

            Assert.True(game.IsFinished);
        }



        [Fact]
        public void GivenA2xA2GridWith2Cells_GameIsFinishedAfterTwoRounds()
        {
            var cell1 = new Cell(1,1);
            var cell2 = new Cell(1,1);
            var grid = Grid.TryCreate(2, 2, new List<Cell> { cell1,cell2 });

            var game = new Game(grid);
            game.Run();
            game.Run();
            Assert.True(game.IsFinished);

        }

        [Fact]
        public void GivenA2xA2GridWith3Cells_GameIsFinishedAfterTwoRounds()
        {
            var cell1 = new Cell(1,1);
            var cell2 = new Cell(1,2);
            var cell3 = new Cell(2,1);
            var grid = Grid.TryCreate(2, 2, new List<Cell> { cell1, cell2, cell3 });

            var game = new Game(grid);
            game.Run();
            game.Run();
            Assert.True(game.IsFinished);
            Assert.False(game.GridIsEmpty());

        }


        [Fact]
        public void GivenABlockInA4x4Grid_GameIsFinishedAfterOneRound()
        {
            // Arrange
            var cell1 = new Cell(2, 2);
            var cell2 = new Cell(2, 3);
            var cell3 = new Cell(3, 2);
            var cell4 = new Cell(3, 3);
            var grid = Grid.TryCreate(4, 4, new List<Cell> { cell1, cell2, cell3, cell4 });
            var game = new Game(grid);

            // Act
            game.Run();

            // Assert
            Assert.True(game.IsFinished);
        }

        [Fact]
        public void Given3CellsInA2x3Grid_GameIsNotFinishedAfterOneRound()
        {
            // Arrange
            var cell1 = new Cell(1, 1);
            var cell2 = new Cell(2, 2);
            var cell3 = new Cell(1, 3);
            var grid = Grid.TryCreate(2, 3, new List<Cell> { cell1, cell2, cell3 });
            var game = new Game(grid);

            // Act
            game.Run();

            // Assert
            Assert.False(game.IsFinished);
        }


        [Fact]
        public void Given6CellsInA4x3Grid_GameIsNotFinishedAfterOneRound()
        {
            // Arrange
            var cell1 = new Cell(2, 1);
            var cell2 = new Cell(3, 1);
            var cell3 = new Cell(2, 2);
            var cell4 = new Cell(3, 2);
            var cell5 = new Cell(2, 3);
            var cell6 = new Cell(3, 3);
            var grid = Grid.TryCreate(4, 3, new List<Cell> { cell1, cell2, cell3, cell4, cell5, cell6 });
            var game = new Game(grid);

            // Act
            game.Run();

            // Assert
            Assert.False(game.IsFinished);

        }
    }
}

File: cpp/01-commands/explain/LifeGame/GameTest.cpp
#include "pch.h"
#include "CppUnitTest.h"
#include "Game.h"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace LifeGame
{
	TEST_CLASS(GameTest)
	{
	public:
		
		TEST_METHOD(WhenGrisIsEmpty_GameFinishesAtNextRound)
		{
            // Arrange
			Grid grid(1, 1, { });
            Game game(std::move(grid));

            // Act
            game.Run();

            // Assert
            Assert::IsTrue(game.IsFinished());
		}

		TEST_METHOD(WhenGridIsInitialized_GameIsNotFinished)
		{
            // Arrange
            Grid grid(1, 1, { });
            Game game(std::move(grid));

            // Assert
            Assert::IsFalse(game.IsFinished());
        }

        TEST_METHOD(WhenGrid1x1With1Cell_GameContinuesAfter1Round)
        {
            // Arrange
            Grid grid(1, 1, { Cell(1, 1) });
            Game game(std::move(grid));

            // Act
            game.Run();

            // Assert
            Assert::IsFalse(game.IsFinished());
        }

        TEST_METHOD(WhenGrid1x1With1Cell_GameFinishesAfter2Rounds)
        {
            // Arrange
            Grid grid(1, 1, { Cell(1, 1) });
            Game game(std::move(grid));

            // Act
            game.Run();
            game.Run();

            // Assert
            Assert::IsTrue(game.IsFinished());
        }

        TEST_METHOD(WhenGrid2x2With2Cells_GameFinishesAfter2Rounds)
        {
            // Arrange
            Grid grid(2, 2, { Cell(1, 1), Cell(1, 2) });
            Game game(std::move(grid));

            // Act
            game.Run();
            game.Run();

            // Assert
            Assert::IsTrue(game.IsFinished());
        }

        TEST_METHOD(WhenGrid2x2With3Cells_GameFinishesAfter2Rounds)
        {
            // Arrange
            Grid grid(2, 2, { Cell(1, 1), Cell(1, 2), Cell(2, 1) });
            Game game(std::move(grid));

            // Act
            game.Run();
            game.Run();

            // Assert
            Assert::IsTrue(game.IsFinished());
        }

        TEST_METHOD(WhenGrid4x4WithCellsBlock_GameFinishesAfter1Round)
        {
            // Arrange
            Grid grid(4, 4, { Cell(2, 2), Cell(2, 3), Cell(3, 2), Cell(3, 3) });
            Game game(std::move(grid));

            // Act
            game.Run();

            // Assert
            Assert::IsTrue(game.IsFinished());
        }

        TEST_METHOD(WhenGrid2x3With3Cells_GameContinuesAfter1Round)
        {
            // Arrange
            Grid grid(2, 3, { Cell(1, 1), Cell(2, 2), Cell(2, 3) });
            Game game(std::move(grid));

            // Act
            game.Run();

            // Assert
            Assert::IsFalse(game.IsFinished());
        }

        TEST_METHOD(WhenGrid4x3With6Cells_GameContinuesAfter1Round)
        {
            // Arrange
            Grid grid(4, 3, { Cell(2, 1), Cell(3, 1), Cell(2, 2), Cell(3, 2), Cell(2, 3), Cell(3, 3) });
            Game game(std::move(grid));

            // Act
            game.Run();

            // Assert
            Assert::IsFalse(game.IsFinished());
        }
    };
}
